File: video/src/shorts.ts
import type { Lesson, LessonCue } from '../../app/src/lib/types';

/**
 * One card of a unit, as a short.
 *
 * The roadmap says `shorts.py` should walk `guide.ts` card-by-card. It walks
 * the *cue list* instead, and the difference is the whole reason this is cheap:
 * a cue records the second its narration starts, so a card already has an in
 * and an out inside the unit's MP3. Nothing is re-synthesised and no audio is
 * cut — Remotion trims the file it plays. A short costs the frames it takes to
 * draw and nothing else.
 *
 * The cost of reading it this way is the same one the app already lives with: a
 * card added to a guide after its unit was narrated has no cue, so it has no
 * short until the unit is re-rendered. The player says "Added since this was
 * recorded" about exactly those cards. This is that property, not a new one.
 */
export interface Short {
  unit: number;
  /** Which card of the unit, counting from 0. */
  card: number;
  question: string;
  answer: string;
  /** Seconds into the unit's MP3 where the question starts. */
  start: number;
  /** Seconds into the unit's MP3 where the answer starts. */
  answerAt: number;
  /** Seconds into the unit's MP3 where the next card begins, or the unit ends. */
  end: number;
}

/** Seconds a short runs for. */
export function shortLength(s: Short): number {
  return s.end - s.start;
}

/**
 * Every card of a lesson, as a short.
 *
 * `pipeline/lessons.py` writes cues as `title`, then a `q`/`a` pair per card,
 * then `close`. A short is one of those pairs plus the silence the renderer
 * leaves after the question — `lesson_script` gives every `q` a two-second
 * pause, which on a lecture is a beat to answer in your head and on a short is
 * the hook holding before the payoff.
 *
 * A `q` with no `a` after it is skipped rather than guessed at. That cannot
 * happen from `lesson_script` today, and a short that showed a question with
 * no answer would be worse than one that does not exist.
 */
export function shortsFor(lesson: Lesson): Short[] {
  const cues: readonly LessonCue[] = lesson.cues ?? [];
  const out: Short[] = [];

  for (let i = 0; i < cues.length; i += 1) {
    if (cues[i].kind !== 'q') continue;
    const answer = cues[i + 1];
    if (!answer || answer.kind !== 'a') continue;

    /*
     * The card ends where the next cue after the answer begins — the next
     * question, or the unit's closing slide. Past the last cue there is only
     * the narration's own end.
     */
    const next = cues[i + 2];
    const end = next ? next.at : lesson.seconds;
    if (end <= cues[i].at) continue;

    out.push({
      unit: lesson.unit,
      card: out.length,
      question: cues[i].text,
      answer: answer.text,
      start: cues[i].at,
      answerAt: answer.at,
      end,
    });
  }

  return out;
}

File: pipeline/shorts.py
#!/usr/bin/env python3
"""
Cut a course's narrated lessons into one vertical short per flashcard.

    python3 pipeline/shorts.py econ --unit 3
    python3 pipeline/shorts.py econ --all --dry-run
    python3 pipeline/shorts.py econ --unit 3 --card 2

Step 2 of docs/VIDEO_PODCAST_ROADMAP.md, and the cheapest format in it: a short
is a question, the beat the narration already leaves after it, and the answer —
cut straight out of the unit's MP3 by the cue list, at 1080x1920.

Nothing is synthesised and no audio file is written. A cue records the second
its line begins, so a card already has an in and an out inside a file every
student streams; Remotion trims what it plays. That is why 278 shorts across
four courses cost compute and nothing else.

The roadmap said this script should walk `guide.ts` card-by-card. It walks the
cue list instead, for the timings — which means a card added to a guide after
its unit was narrated has no short until the unit is re-rendered. The app
already lives with exactly that: the lesson player calls those cards "Added
since this was recorded".

Output goes to app/public/audio/shorts/<course>/unit-<n>-card-<m>.mp4, which is
gitignored along with the rest of the rendered video — see video/README.md.
"""

from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent


def main() -> int:
    ap = argparse.ArgumentParser(
        description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter
    )
    ap.add_argument("course")
    ap.add_argument("--unit", type=int, help="Cut one unit only.")
    ap.add_argument("--card", type=int, help="Cut one card of that unit only.")
    ap.add_argument("--all", action="store_true", help="Cut every unit of the course.")
    ap.add_argument("--ground", help="Ground id from lib/look.ts (default: ink).")
    ap.add_argument("--accent", help="Accent id from lib/look.ts (default: sterling).")
    ap.add_argument(
        "--dry-run",
        action="store_true",
        help="Print the shot list and what it would cost, and render nothing.",
    )
    args = ap.parse_args()

    if args.unit is None and not args.all:
        ap.error("pass --unit N for one unit, or --all for the whole course")

    video = ROOT / "video"
    if not (video / "node_modules").exists():
        print(
            f"{video}/node_modules is missing — run `npm install` in {video} first.\n"
            "It is a separate package on purpose: Remotion brings a browser with it, "
            "and the app's own install should not carry that.",
            file=sys.stderr,
        )
        return 1

    cmd = ["node", "render-shorts.mjs", args.course]
    if args.unit is not None:
        cmd += ["--unit", str(args.unit)]
    if args.card is not None:
        cmd += ["--card", str(args.card)]
    if args.all:
        cmd.append("--all")
    if args.ground:
        cmd += ["--ground", args.ground]
    if args.accent:
        cmd += ["--accent", args.accent]
    if args.dry_run:
        cmd.append("--dry-run")

    return subprocess.run(cmd, cwd=video).returncode


if __name__ == "__main__":
    raise SystemExit(main())
